src/js/characters.ts

```diff
@@ -0,0 +1,378 @@
+
+export class Sprite {
+  readonly name: string;
+  readonly converter: Map<number, number[]>;
+  readonly image: string;
+  readonly chr_data: number[];
+  readonly description?: string;
+  constructor(name: string, converter: Map<number, number[]>, image: string, chr_data: number[], description?: string) {
+    this.name = name;
+    this.converter = converter;
+    this.image = image;
+    this.chr_data = chr_data;
+    this.description = description;
+  }
+
+  public applyPatch(rom: Uint8Array) {
+    if (this.chr_data.length == 0) {
+      return;
+    }
+    for (let [src, dsts] of this.converter) {
+      for (let dst of dsts) {
+        for (let i=0; i<0x10; ++i) {
+          rom[dst + i] = this.chr_data[src * 0x10 + i];
+        }
+      }
+    }
+  }
+}
+
+export class CharacterSet {
+  private static instance: CharacterSet;
+  private readonly semiaReplacements = new Array<Sprite>();
+
+  static semia(): Sprite[] {
+    if (!this.instance) this.instance = new CharacterSet();
+    return [...this.instance.semiaReplacements];
+  }
+
+  constructor() {
+    this.semiaReplacements.push(new Sprite("Semia", CustomTilesetMapping.semia(), "images/semia.png", [], "The original main character of Crystalis"));
+    this.semiaReplacements.push(new Sprite("Mesia", CustomTilesetMapping.semia(), "images/mesia.png", mesia_patch_data, "Secondary protagonist Mesia takes the spotlight! Artwork by jroweboy"));
+  }
+}
+
+function toAddr(chr_page: number, nametable: number, tile_number: number): number {
+  const baseAddr = 0x40000 + 0x10; // added 0x10 to account for rom header
+  return baseAddr + chr_page * 0x2000 + nametable * 0x1000 + tile_number * 0x10;
+}
+
+// Provides a lookup from the sample tileset to the CHRROM locations
+class CustomTilesetMapping {
+  private static instance: CustomTilesetMapping;
+  private readonly semiaMapping: Map<number, number[]>;
+
+  static semia() : Map<number, number[]> {
+    if (!this.instance) this.instance = new CustomTilesetMapping();
+    return this.instance.semiaMapping;
+  }
+
+  constructor() {
+    this.semiaMapping = this.generateSemiaMapping();
+  }
+
+  private generateSemiaMapping() : Map<number, number[]> {
+    // For most of the mappings, there is only one location to write it to, but for some, its split across several CHRROM banks.
+    // so thats why its a map of tileset number to a list of addresses
+
+    // A CHR tileset is 16 tiles wide, the left 8 tiles are the no armor sprites, the right 8 are the armor sprites
+    const ARMOR_TILESET_OFFSET = 0x08;
+    // In the original game, the game swaps the two 0x20 size CHR banks for sprites to switch armor/no armor sprites
+    const CHR_PAGE_OFFSET = 0x400;
+
+    const mapping = new Map<number, number[]>();
```

Comment:
I might prefer to see this as a constructor argument, or some other format.  But this is fine for now.